#VRML_SIM R2023b utf8
PROTO FourWheelsRobot [
  field SFVec3f    translation  0 0 0
  field SFRotation rotation     0 0 1 0
  field SFFloat    bodyMass     1
  field SFString   name         ""
]
{
  Robot {
    translation IS translation
    rotation IS rotation
    name IS name
    children [
        DistanceSensor {
        translation 0.1 0.025 0
        rotation 0 0 1 0.3
        children [
            DEF SENSOR Shape {
            appearance PBRAppearance {
                baseColor 0.0500649 0 1
                roughness 1
                metalness 0
            }
            geometry Box {
                size 0.01 0.01 0.01
            }
            }
        ]
        name "ds_left"
        boundingObject USE SENSOR
        physics DEF SENSOR_PH Physics {
        }
        }
        DistanceSensor {
        translation 0.1 -0.025 0
        rotation 0 0 1 -0.3
        children [
            USE SENSOR
        ]
        name "ds_right"
        boundingObject USE SENSOR
        physics USE SENSOR_PH
        }
        DEF BODY Shape {
        appearance PBRAppearance {
            baseColor 1 0 0
            roughness 1
            metalness 0
        }
        geometry Box {
            size 0.2 0.1 0.05
        }
        }
        HingeJoint {
        jointParameters HingeJointParameters {
            position 19.990560317418716
            axis 0 1 0
            anchor 0.05 0.06 0
        }
        device [
            RotationalMotor {
            name "wheel1"
            }
        ]
        endPoint DEF WHEEL1 Solid {
            translation 0.05 0.06 0
            rotation -0.7405383154562751 -0.47518488218309607 0.4751866276378359 4.41642529735683
            children [
            DEF WHEEL Shape {
                appearance PBRAppearance {
                baseColor 0.100618 1 0
                roughness 1
                metalness 0
                }
                geometry Cylinder {
                height 0.02
                radius 0.04
                }
            }
            ]
            name "wheel1"
            boundingObject USE WHEEL
            physics DEF WHEEL_PH Physics {
            }
        }
        }
        HingeJoint {
        jointParameters HingeJointParameters {
            position -2.2059311292083911e-07
            axis 0 1 0
            anchor 0.05 -0.06 0
        }
        device [
            RotationalMotor {
            name "wheel2"
            }
        ]
        endPoint DEF WHEEL2 Solid {
            translation 0.05 -0.06 0
            rotation 0.9999999999999878 -1.1037328966792166e-07 1.1037369509239713e-07 1.570800000000012
            children [
            USE WHEEL
            ]
            name "wheel2"
            boundingObject USE WHEEL
            physics USE WHEEL_PH
        }
        }
        HingeJoint {
        jointParameters HingeJointParameters {
            position 1.3731348979945077e-07
            axis 0 1 0
            anchor -0.05 0.06 0
        }
        device [
            RotationalMotor {
            name "wheel3"
            }
        ]
        endPoint DEF WHEEL3 Solid {
            translation -0.05 0.06 0
            rotation 0.9999999999999953 6.859394668222963e-08 -6.859419864232737e-08 1.5708000000000044
            children [
            USE WHEEL
            ]
            name "wheel3"
            boundingObject USE WHEEL
            physics USE WHEEL_PH
        }
        }
        HingeJoint {
        jointParameters HingeJointParameters {
            position -3.9991499356228204e-08
            axis 0 1 0
            anchor -0.05 -0.05 0
        }
        device [
            RotationalMotor {
            name "wheel4"
            }
        ]
        endPoint DEF WHEEL4 Solid {
            translation -0.05 -0.06 0
            rotation 0.9999999999999996 -1.9995676229759237e-08 1.9995749678114102e-08 1.5708
            children [
            USE WHEEL
            ]
            name "wheel4"
            boundingObject USE WHEEL
            physics USE WHEEL_PH
        }
        }
    ]
    boundingObject USE BODY
    physics Physics {
      density -1
      mass IS bodyMass
    }
    controller "four_wheeled_collision_avoidance"
	}
}